### PR TITLE
feat(origin detection): implement External Data Env

### PR DESCRIPTION
**Pull request by wdhif**
### What does this PR do?

Implement External Data environment variable injection in the Admission Controller. Format for this new environment variable would be `it-INIT_CONTAINER,cn-CONTAINER_NAME,pu-POD_UID` and would look like:

```
it-false,cn-nginx-webserver,pu-75a2b6d5-3949-4afb-ad0d-92ff0674e759
```

### Motivation

This new variable is needed for the New Origin Detection spec, especially as a solution to resolve Origin Detection in case Local Data are unavailable, for example with Kata Containers and CGroupv2.

### Describe how to test/QA your changes

* Deploy the Agent and dummy workload like those:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
  annotations:
    environment: testing
    user: wassim
  labels:
    environment: testing
    user: wassim
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      annotations:
        environment: testing
        user: wassim
      labels:
        app: nginx
        admission.datadoghq.com/enabled: "true"
        environment: testing
        user: wassim
    spec:
      containers:
      - name: nginx
        image: nginx
        env:
        - name: DD_AGENT_HOST
          valueFrom:
            fieldRef:
              fieldPath: status.hostIP
```

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: init
spec:
  replicas: 1
  selector:
    matchLabels:
      app: init
  template:
    metadata:
      annotations:
        environment: testing
        user: wassim
      labels:
        app: init
        admission.datadoghq.com/enabled: "true"
        environment: testing
        user: wassim
    spec:
      initContainers:
      - name: init
        image: busybox
        command: ["sh", "-c", "while true; do sleep 3600; done"]
      containers:
      - name: nginx
        image: nginx
```
Note that you need the `admission.datadoghq.com/enabled` annotation.
* Verify the environment variables assigned to the workloads
```
➜  datadog-dev git:(main) ✗ k exec -it deployments/nginx -- env | grep DD_
DD_TRACE_AGENT_URL=unix:///var/run/datadog/apm.socket
DD_AGENT_HOST=192.168.49.2
DD_INTERNAL_POD_UID=9b00b901-ccf9-4a02-b35f-4b718c34b2c6
DD_EXTERNAL_ENV=it-false,cn-nginx,pu-9b00b901-ccf9-4a02-b35f-4b718c34b2c6
DD_INSTRUMENTATION_INSTALL_ID=d517565f-1a32-426a-87de-113335a9eeca
DD_INSTRUMENTATION_INSTALL_TIME=1720102315
DD_ENTITY_ID=9b00b901-ccf9-4a02-b35f-4b718c34b2c6
DD_DOGSTATSD_URL=unix:///var/run/datadog/dsd.socket
```
```
➜  datadog-dev git:(main) ✗ k exec -it deployments/init -c init -- env | grep DD_
DD_ENTITY_ID=b6eec695-76c2-4a7c-96b1-90d914c45665
DD_DOGSTATSD_URL=unix:///var/run/datadog/dsd.socket
DD_TRACE_AGENT_URL=unix:///var/run/datadog/apm.socket
DD_INTERNAL_POD_UID=b6eec695-76c2-4a7c-96b1-90d914c45665
DD_EXTERNAL_ENV=it-true,cn-init,pu-b6eec695-76c2-4a7c-96b1-90d914c45665
DD_INSTRUMENTATION_INSTALL_ID=14acf80d-2b7b-4aae-9764-431dc3c07ab5
DD_INSTRUMENTATION_INSTALL_TIME=1720195544
```

### CODE REVIEW

**Review by domalessi — APPROVED** [review]

File: pkg/clusteragent/admission/mutate/config/config.go [hunk 1]
```diff
@@ -163,19 +171,22 @@ func (w *Webhook) mutate(request *admission.MutateRequest) ([]byte, error) {
...
 func (w *Webhook) inject(pod *corev1.Pod, _ string, _ dynamic.Interface) (bool, error) {
-	var injectedConfig, injectedEntity bool
+	var injectedConfig, injectedEntity, injectedExternalEnv bool
 
 	if pod == nil {
 		return false, errors.New(metrics.InvalidInput)
-
 	}
 
 	if !autoinstrumentation.ShouldInject(pod, w.wmeta) {
```
Comment (adel121) [inline pkg/clusteragent/admission/mutate/config/config.go:185]: just confirming, if autoinstrumentation is not enabled for the current pod, should we not inject DD_EXTERNAL_ENV?

File: pkg/clusteragent/admission/mutate/config/config.go [hunk 2]
```diff
@@ -217,6 +232,53 @@ func injectionMode(pod *corev1.Pod, globalMode string) string {
 	return globalMode
 }
 
+// injectExternalDataEnvVar injects the External Data environment variable.
+// The format is: it-<init>,cn-<container_name>,pu-<pod_uid>
+func injectExternalDataEnvVar(pod *corev1.Pod) bool {
```
Comment (adel121) [inline pkg/clusteragent/admission/mutate/config/config.go:237]: nit: you can use named returns in golang to make it more obvious what the returned boolean signifies

File: pkg/clusteragent/admission/mutate/config/config.go [hunk 3]
```diff
@@ -217,6 +232,53 @@ func injectionMode(pod *corev1.Pod, globalMode string) string {
...
+	}
+	var containerInjections []containerInjection
+
+	// Collect all containers and init containers
+	for i := range pod.Spec.Containers {
+		containerInjections = append(containerInjections, containerInjection{&pod.Spec.Containers[i], false})
+	}
+	for i := range pod.Spec.InitContainers {
+		containerInjections = append(containerInjections, containerInjection{&pod.Spec.InitContainers[i], true})
+	}
```
Comment (adel121) [inline pkg/clusteragent/admission/mutate/config/config.go:252]: not sure if we can avoid this overhead, especially that we are in a webhook. 

The alternative I am thinking about is having a helper function like this:

```
func injectExternalDataEnvVarToContainersList(ctrs []*corev1.Container, init bool) {
// Loop directly on ctrs and inject the env var
}
```

After that you can simplify the function you wrote into something like:
```
injectExternalDataEnvVarToContainersList(pod.Spec.Containers, false)
injectExternalDataEnvVarToContainersList(pod.Spec.InitContainers, true)
```

I guess this makes the code easier to read and debug, and also is slightly better in terms of performance (no need to define the new struct type and copy containers into slices)

WDYT?

**Review by adel121 — COMMENTED** [review]
thanks @wdhif 

added few comments 👍

File: pkg/clusteragent/admission/mutate/config/config.go [hunk 4]
```diff
@@ -192,13 +203,17 @@ func (w *Webhook) inject(pod *corev1.Pod, _ string, _ dynamic.Interface) (bool,
...
+	// Inject DD_ENTITY_ID
 	if w.config.injectContName {
 		injectedEntity = injectFullIdentity(pod)
 	} else {
 		injectedEntity = common.InjectEnv(pod, defaultDdEntityIDEnvVar)
 	}
 
-	return injectedConfig || injectedEntity, nil
+	// Inject External Data Environment Variable
+	injectedExternalEnv = injectExternalDataEnvVar(pod)
```
[the remark this answers is not in the document]
Comment (wdhif) [reply pkg/clusteragent/admission/mutate/config/config.go:214]: Yes in this case it's fine.

File: pkg/clusteragent/admission/mutate/config/config.go (hunk 2 above)
Comment (wdhif) [reply pkg/clusteragent/admission/mutate/config/config.go]: Done 👍 I've not used a naked return though to maintain readability.

File: pkg/clusteragent/admission/mutate/config/config.go (hunk 3 above)
Comment (wdhif) [reply pkg/clusteragent/admission/mutate/config/config.go:250]: Checked together, current implementation ended up being the best option.

**Review by adel121 — APPROVED** [review]

**Pull request merged**